x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/logging/Throttler.java

```diff
@@ -0,0 +1,163 @@
+ * If a log call occurs after the throttle period, then the call will proceed, and it will include a message like
+ * "repeated X times" to indicate how often the message was attempting to be logged.
+ */
+public class Throttler implements Closeable {
+
+    private static final Logger classLogger = LogManager.getLogger(Throttler.class);
+
+    private final TimeValue resetInterval;
+    private Duration durationToWait;
+    private final Clock clock;
+    private final ConcurrentMap<String, LogExecutor> logExecutors;
+    private final AtomicReference<Scheduler.Cancellable> cancellableTask = new AtomicReference<>();
+
+    /**
+     * Constructs the throttler and kicks of a scheduled tasks to clear the internal stats.
+     *
+     * @param resetInterval the frequency for clearing the internal stats. This protects against an ever growing
+     *                      cache
+     * @param durationToWait the amount of time to wait before logging a message after the threshold
+     *                       is reached
+     * @param threadPool a thread pool for running a scheduled task to clear the internal stats
+     */
+    public Throttler(TimeValue resetInterval, TimeValue durationToWait, ThreadPool threadPool) {
+        this(resetInterval, durationToWait, Clock.systemUTC(), threadPool, new ConcurrentHashMap<>());
+    }
+
+    /**
+     * This should only be used directly for testing.
+     */
+    Throttler(
+        TimeValue resetInterval,
+        TimeValue durationToWait,
+        Clock clock,
+        ThreadPool threadPool,
+        ConcurrentMap<String, LogExecutor> logExecutors
+    ) {
+        Objects.requireNonNull(durationToWait);
+        Objects.requireNonNull(threadPool);
+
+        this.resetInterval = Objects.requireNonNull(resetInterval);
+        this.durationToWait = Duration.ofMillis(durationToWait.millis());
+        this.clock = Objects.requireNonNull(clock);
+        this.logExecutors = Objects.requireNonNull(logExecutors);
+
+        this.cancellableTask.set(startResetTask(threadPool));
+    }
+
+    private Scheduler.Cancellable startResetTask(ThreadPool threadPool) {
+        classLogger.debug(() -> format("Reset task scheduled with interval [%s]", resetInterval));
+
+        return threadPool.scheduleWithFixedDelay(logExecutors::clear, resetInterval, threadPool.executor(UTILITY_THREAD_POOL_NAME));
+    }
+
+    public void setDurationToWait(TimeValue durationToWait) {
+        this.durationToWait = Duration.ofMillis(durationToWait.millis());
+    }
+
+    public void warn(Logger logger, String message, Throwable e) {
+        Objects.requireNonNull(message);
+        Objects.requireNonNull(e);
+
+        logHelper(message, msgToAppend -> logger.warn(message.concat(msgToAppend), e));
+    }
+
+    private void logHelper(String message, Consumer<String> executor) {
+        LogExecutor logExecutor = logExecutors.compute(message, (key, value) -> {
```

Comment:
My understanding is this is an atomic call, so two calls should get separate `LogExecutor` objects returned by the `compute` call. I believe this will block other `compute` calls.